src/main/host/syscall/handler/signal.rs

```diff
@@ -1,20 +1,82 @@
 
 impl SyscallHandler {
     #[log_syscall(/* rv */ std::ffi::c_int, /* pid */ linux_api::posix_types::kernel_pid_t,
                   /* sig */ std::ffi::c_int)]
     pub fn kill(
         ctx: &mut SyscallContext,
-        _pid: linux_api::posix_types::kernel_pid_t,
-        _sig: std::ffi::c_int,
+        pid: linux_api::posix_types::kernel_pid_t,
+        sig: std::ffi::c_int,
     ) -> Result<(), SyscallError> {
-        let rv = Self::legacy_syscall(c::syscallhandler_kill, ctx)?;
-        assert_eq!(0, i32::from(rv));
+        log::trace!("kill called on pid {pid} with signal {sig}");
+
+        let pid = if pid == -1 {
+            // kill(2): If pid equals -1, then sig is sent to every process for which the calling
+            // process has permission to send signals, except for process 1.
+            //
+            // Currently unimplemented, and unlikely to be needed in the context of a shadow
+            // simulation.
+            log::warn!("kill with pid=-1 unimplemented");
+            return Err(Errno::ENOTSUP.into());
+        } else if pid == 0 {
+            // kill(2): If pid equals 0, then sig is sent to every process in the process group of
+            // the calling process.
+            //
+            // Currently every emulated process is in its own process group.
+            ctx.objs.process.id()
+        } else if pid < -1 {
+            // kill(2): If pid is less than -1, then sig is sent to every process in the process
+            // group whose ID is -pid.
+            //
+            // Currently every emulated process is in its own process group, where pgid=pid.
+            (-pid).try_into().or(Err(Errno::ESRCH))?
+        } else {
+            pid.try_into().or(Err(Errno::ESRCH))?
+        };
+
+        let Some(target_process) = ctx.objs.host.process_borrow(pid) else {
+            log::debug!("Process {pid} not found");
+            return Err(Errno::ESRCH.into());
+        };
+        let target_process = &*target_process.borrow(ctx.objs.host.root());
+
+        Self::signal_process(ctx.objs, target_process, sig)
+    }
+
+    /// Send a signal to `target_process` from the thread and process in `objs`. A signal of 0 will
+    /// be ignored.
+    fn signal_process(
+        objs: &ThreadContext,
+        target_process: &Process,
+        signal: std::ffi::c_int,
+    ) -> Result<(), SyscallError> {
+        if signal == 0 {
+            return Ok(());
+        }
+
+        let Ok(signal) = Signal::try_from(signal) else {
+            return Err(Errno::EINVAL.into());
+        };
+
+        if signal.is_realtime() {
+            log::warn!("Unimplemented signal {signal:?}");
+            return Err(Errno::ENOTSUP.into());
+        }
+
+        let sender_pid = objs.process.id().into();
+        let siginfo = unsafe { siginfo_t::peel(siginfo_t::new_for_kill(signal, sender_pid, 0)) };
+        let siginfo = unsafe { siginfo_t::wrap_ref_assume_initd(&siginfo) };
```

Comment:
I realized that this can be simplified to just the following, so I've also made this change:

```rust
let siginfo = siginfo_t::new_for_kill(signal, sender_pid, 0);
```